Comment like models

wathigo opened the pull request:
Task Description:
1. Creation of Like and Comment models.
2. Implementation of Like and comments controllers and views
3. Implementation of integration tests.
Please review my PR. Thank you.

mcrd25 commented:
# Currently under review by @mcrd25

mcrd25 commented on fb/app/models/post.rb:
```diff
@@ -1,4 +1,8 @@
-  belongs_to :creator,       class_name: 'User'
+  belongs_to :creator, class_name: 'User'
+  has_many :likes, as: :likeable
+  has_many :comments, dependent: :destroy
```
Good job adding `dependent: :destroy`

wathigo replied:
Thank you

mcrd25 commented on fb/app/controllers/comments_controller.rb:
```diff
@@ -0,0 +1,27 @@
+
+  def destroy
+    @comment = Comment.find_by_id(params[:format])
+    if current_user.id == @comment.user_id
```
You should check if the `@comment` even exists first [OPTIONAL]

wathigo replied:
Sure, I will handle this edge case.

mcrd25 commented on fb/app/controllers/likes_controller.rb:
```diff
@@ -0,0 +1,31 @@
+
+  def destroy
+    @liked ||= likeable.likes.find_by(user_id: current_user.id)
+    @liked.delete
```
What is delete of like was unsuccessful? [OPTIONAL]

wathigo replied:
I will also handle this edge case too

wathigo commented:
Thank you @mcrd25 for reviewing my PR. I will consider using flash to render user friendly messages on failed operations. I have implemented the required changes.

SunnyAsar commented on fb/app/controllers/comments_controller.rb:
```diff
@@ -0,0 +1,29 @@
+    if @comment.save
+      redirect_back(fallback_location: root_path)
+    else
+      render :new
```
submitting a blank comment breaks the app:
- remember comments is embedded on that page
- also comments can become a lot, lets add a link on the post to go to a show page
![Screenshot 2019-09-10 at 5 31 28 PM](https://user-images.githubusercontent.com/10922508/64632492-0ff7fc00-d3f1-11e9-87ef-74879f6a02af.png)

mcrd25 replied:
Good catch 🙂

wathigo replied:
Thanks Sunny. This definitely needs to be fixed. I will implement the changes then submit another request.

wathigo commented:
Hello @SunnyAsar . I have implemented the following changes according to your review.
1. There is a problem with what we try to render when a comment isn't made. 
I have changed this so that nothing happens when you submit an empty comment

2. Lets add a show page in case where comments becomes a lot. 
 I had not thought about this case during my original implementation so creating a comments show page would require me to start a new design. However, I have written some Js code that allows user to see all the comments while still on the home page.

3. We have minor linting problems, please make it all green.
I have fixed most of the linting errors. There are still two errors where one resides in the schema.rb and I don't think it's a good idea to interfere with this file.
The second error requires refactoring but am find it hard to refactor according to the linter recommendations and still maintain code semantics

Please review the changes. Thank you

SunnyAsar commented on fb/app/controllers/comments_controller.rb:
```diff
@@ -0,0 +1,23 @@
+
+  def create
+    @comment = current_user.comments.build(comment_params)
+    redirect_back(fallback_location: root_path) if @comment.save
```
This solution isn't the best either. The form will be locked once the first attempt is blank and since its not re rendering, you can comment till you refresh. which is not good.
**Consider this**

```
def create
    @comment = current_user.comments.build(comment_params)
    if @comment.save
      redirect_to root_path
    else
      @pagy, @posts = pagy_array(Post.all.sort_by(&:created_at).reverse)  // you need this line so as to send the posts on that page back to be rendered.
      flash.now[:alert] = 'no blank comments'  // flash.now is a alternative that makes flash available instantly without a new request cycle.
      render 'posts/index'
    end
  end

```

- instead of using flash, consider returning form errors the way you did for the create post form.

wathigo replied:
Sure, I will fix this one.

wathigo commented:
Thank you @SunnyAsar . I thought you were referring to comments #show. I will definitely add post #show in my next milestone.
